rust/ql/lib/codeql/rust/internal/TypeInference.qll

```diff
@@ -818,1663 +791,2190 @@ private Type inferPathExprType(PathExpr pe, TypePath path) {
+   *
+   * `blanketPath` points to the type `blanketTypeParam` inside `selfType`, which
+   * is the type parameter used in the blanket implementation.
+   *
+   * This predicate only checks for matching method names and arities, and whether
+   * the trait being implemented by `i` (when `i` is not a trait itself) is visible
+   * at `mc`.
+   */
+  bindingset[mc]
+  pragma[inline_late]
+  private predicate methodCallBlanketCandidate(
+    MethodCall mc, Method m, ImplItemNode impl, AssocFunctionType self, TypePath blanketPath,
+    TypeParam blanketTypeParam
+  ) {
+    exists(string name, int arity |
+      mc.hasNameAndArity(name, arity) and
+      methodInfoBlanket(m, name, arity, impl, _, self, blanketPath, blanketTypeParam)
+    |
+      methodCallVisibleImplTraitCandidate(mc, impl)
       or
-      dpos.isSelf() and
-      result = this.getReturnType(path)
-    }
+      impl.resolveTraitTy() = mc.(Call).getTrait()
+    )
   }
 
-  private class TupleStructDecl extends TupleDeclaration, Struct {
-    TupleStructDecl() { this.isTuple() }
+  /**
+   * A method call.
+   *
+   * This is either:
+   *
+   * 1. `MethodCallMethodCallExpr`: an actual method call, `x.m()`;
+   * 2. `MethodCallIndexExpr`: an index expression, `x[i]`, which is [syntactic sugar][1]
+   *    for `*x.index(i)`;
+   * 3. `MethodCallCallExpr`: a qualified function call, `Q::m(x)`, where `m` is a method;
+   *    or
+   * 4. `MethodCallOperation`: an operation expression, `x + y`, which is syntactic sugar
+   *    for `Add::add(x, y)`.
+   *
+   * Note that only in case 1 and 2 is auto-dereferencing and borrowing allowed.
+   *
+   * [1]: https://doc.rust-lang.org/std/ops/trait.Index.html
+   */
+  abstract class MethodCall extends Expr {
+    abstract predicate hasNameAndArity(string name, int arity);
 
-    override TypeParameter getTypeParameter(TypeParameterPosition ppos) {
-      typeParamMatchPosition(this.getGenericParamList().getATypeParam(), result, ppos)
+    abstract Expr getArgument(ArgumentPosition pos);
+
+    abstract predicate supportsAutoDerefAndBorrow();
+
+    AstNode getNodeAt(FunctionPosition apos) {
+      result = this.getArgument(apos.asArgumentPosition())
+      or
+      result = this and apos.isReturn()
     }
 
-    override Type getParameterType(DeclarationPosition dpos, TypePath path) {
-      exists(int pos |
-        result = this.getTupleField(pos).getTypeRepr().(TypeMention).resolveTypeAt(path) and
-        pos = dpos.asPosition()
+    Type getArgumentTypeAt(ArgumentPosition pos, TypePath path) {
+      result = inferType(this.getArgument(pos), path)
+    }
+
+    private Type getReceiverTypeAt(TypePath path) {
+      result = this.getArgumentTypeAt(any(ArgumentPosition pos | pos.isSelf()), path)
+    }
+
+    /**
+     * Same as `getACandidateReceiverTypeAt`, but without borrows.
+     */
+    pragma[nomagic]
+    private Type getACandidateReceiverTypeAtNoBorrow(TypePath path, string derefChain) {
```

Comment:
I would prefer to put the path parameter last as we seem to have a convention where predicates that return a type take the path at the end. `derefChain` is also the more stable argument, so it feels natural to have it first.

If you agree, then there's a few other predicates where the same comment applies. The regex `TypePath.*, string derefChain` finds most (all?) of them.